### ISSUE TITLE
Increase concurrency between script and layout

### ISSUE THREAD

**Issue by asajeffrey** [servo/servo#12513]
Implement a finer-grained locking strategy (or preferably lock-free) to allow script to run in parallel with layout more than it currently does. cc @metajack

**Comment by asajeffrey** [comment]
IRC conversation with @pcwalton http://logs.glob.uno/?c=mozilla%23servo&s=19+Jul+2016&e=19+Jul+2016#c484008

@jdm points out https://html.spec.whatwg.org/multipage/webappapis.html#event-loop-processing-model

**Comment by asajeffrey** [comment]
OK, a simple model...

Add a reentrant lock to each script thread.

Script grabs the lock the first time it calls a DOM method, and releases it at the end of the end of the turn of the event loop.

Layout grabs the lock when it starts one run of style/layout, and releases it at the end.

The extra concurrency would be that JS code can run concurrently with layout up until it first touches the DOM. For code using a framework, this may be significant, e.g. model update, computing VDOM, calculating VDOM diffs, can all be done concurrently, it's only the final DOM mutation that needs the lock.

cc @nox

**Comment by asajeffrey** [comment]
As a not-so-simple model, we could divide (e.g. in WebIDL) methods that are visible to layout from those that aren't, and only gab the lock when calling methods that are visible to layout. We'd have to make sure that methods that are layout-invisible don't do anything layout-visible.

Variants of this include using a read-write lock (so layout can run concurrently with reading the DOM, just not modifying it), passing round access tokens, etc.

**Comment by asajeffrey** [comment]
Discussion on #servo https://matrix.to/#/!crXxRMedJXFPxKffWI:mozilla.org/$gqPSkOVbeSHQnmfk3foLjKxiJs-958tbOhRmE1MDVRg?via=mozilla.org&via=matrix.org&via=igalia.com

**Comment by nox** [comment]
> methods that are visible to layout from those that aren't

That's not the right distinction here. Layout doesn't care about WebIDL methods, it cares about data. We would need to introduce new refcell types that need the lock, and use that for all the data that layout accesses. It's doable but that seems an extremely large change to me.

**Comment by asajeffrey** [comment]
@nox I think the simplest thing to do would be to implement the simple model (everything locks) and look to see what the bottlenecks are. One useful data point would be to find out which DOM APIs are used by frameworks during VDOM building and diffing. Rather annoyingly I bet that there is at least one, which is `DOMString`. There are probably others.

**Comment by nox** [comment]
Again, this is not about APIs but about the data used by those APIs.

`DOMString` is not a DOM class.

**Comment by asajeffrey** [comment]
So somewhere there needs to be a dynamic check / lock grab / something to stop layout from seeing inconsistent views of the DOM. It seems like the options are:

- every time a `DOMRefCell` is borrowed
- every time a `DOMRefCell` is borrowed mutably
- every time we enter Rust code from SM
- er something else I've not thought of

@nox, it seems like you are arguing for one of the first two? 

(And yes, `DOMString` is not a DOM class, but it is implemented in Rust, which was the point I was trying to make.)

**Comment by asajeffrey** [comment]
Hmm, if we introduced a `DOMLayoutRefCell<T>` which is the same as `DOMRefCell` but allows layout access as well as script, then I wonder how much work would be needed to ensure that everything layout uses is a `DOMLayoutRefCell`? @nox does layout-2020 distinguish between ref cells layout can use vs ones which are only for script?

**Comment by nox** [comment]
`DOMString` is just a string on the Rust side. I don't understand the point you are trying to make related to that.

Layout 2020 uses the same stuff as layout 2013 to talk to script, I still wish I had time to switch that stuff to inert but AFAIK I don't have it.

To do what you want, we need something like `DomLayoutRefCell<T>` that works just like `DomRefCell<T>` but takes a token to the global lock you want, and make it block until the lock is free.

**Comment by nox** [comment]
Basically, you need to make, say, `Element::attrs` not be accessible from styling or layout when script is running.

**Comment by nox** [comment]
And on the opposite side, make access to `Element::attrs` block from script when styling or layout are running.

**Comment by asajeffrey** [comment]
@nox Read access on both sides is fine, it's write access that we need to worry about? So we could do some kind of read-write locking at the point when script calls `borrow_mut`?

**Comment by nox** [comment]
Yes, sorry, I should have said when script wants to mutate stuff, not just access.

**Comment by gterzian** [comment]
It's not completely clear to me what data you're talking about, and how it is shared, however assuming it's simply a `DomRefCell` that needs to be made threadsafe I think you can try adding a mutex and a condvar per refcell as opposed to a global lock.

It seems [`parking_lot`](https://docs.rs/parking_lot/0.7.0/parking_lot/struct.Condvar.html) was designed around the idea of making it cheap to have a huge number of locks/condvars, and to make it cheap to do uncontended operations on those, which would seem to fit the use case. 

You could try something like the below, which could actually allow script and layout to not block each other unless they would want to write on the same Dom node(perhaps this doesn't make any sense, if for example writing one node could invalidate many others?).

```rust
#[derive(Clone, Debug, Default, PartialEq)]
pub struct DomNodeState {
    // At most 2?
    reader_count: usize,
    readable: bool,
    writable: bool,
    layout_wants_to_write: bool,
}

#[derive(Clone, Debug, Default, MallocSizeOf, PartialEq)]
pub struct DomRefCell<T> {
    value: RefCell<T>,
    #[ignore_malloc_size_of = "Arc"]
    state: Arc<(Mutex<DomState>, Condvar)>, 
}

impl DomRefCell {
    pub fn borrow_mut_for_layout(&self) -> LayoutWrappedRefMut<T> {
        {
            let &(ref lock, ref cvar) = &*self.state; 
            let mut state = lock.lock();
            // A flag like this could be used to give 
            // a limited kind of priority to layout.
            state.layout_wants_to_write = true;
            while !(state.writeable) {
                cvar.wait(&mut state);
            }
            state.readable = false;
            state.writeable = false;
        }  
        debug_assert!(thread_state::get().is_layout());
        let ref_mut = self.value.borrow_mut();
        WrappedRefMut::new(ref_mut, self.state.clone())
    }

    pub fn borrow(&self) -> WrappedRef<T> {
        {
            let &(ref lock, ref cvar) = &*self.state; 
            let mut state = lock.lock();
            while !(state.readable && !state.layout_wants_to_write) {
                cvar.wait(&mut state);
            }
            state.writable = false;
            state.reader_counter += 1;
        }  
        let ref = self.try_borrow()
            .expect("DomRefCell<T> already mutably borrowed");
        WrappedRef::new(ref, self.sate.clone())
    }

    pub fn borrow_mut(&self) -> WrappedRefMut<T> {
        {
            let &(ref lock, ref cvar) = &*self.state; 
            let mut state = lock.lock();
            while !(state.writeable && !state.layout_wants_to_write) {
                cvar.wait(&mut state);
            }
            state.readable = false;
            state.writeable = false;
        }  
        let ref_mut = self.try_borrow_mut()
            .expect("DomRefCell<T> already borrowed");
        WrappedRefMut::new(ref_mut, self.state.clone())
    }
}

impl Drop for WrappedRef {
    fn drop(&mut self) {
        let &(ref lock, ref cvar) = &*self.state; 
        let mut state = lock.lock();
        state.reader_counter -= 1;
        if state.reader_counter  == 0 {
            state.writeable = true;
            cvar.notify_one();
        }
    }
}

impl Drop for WrappedRefMut {
    fn drop(&mut self) {
        let &(ref lock, ref cvar) = &*self.state; 
        state.readable = true;
        state.writeable = true;
        cvar.notify_one();
    }
}

impl Drop for LayoutWrappedRefMut {
    fn drop(&mut self) {
        let &(ref lock, ref cvar) = &*self.state; 
        state.readable = true;
        state.writeable = true;
        state.layout_wants_to_write = false;
        cvar.notify_one();
    }
}


```

**Comment by nox** [comment]
Nah the idea is to use a global lock because we don't want layout and script to run together. Alan just wants script to be able to continue to run until it ever try to access layout stuff. I think this is better than locks everywhere too.

**Comment by gterzian** [comment]
> Alan just wants script to be able to continue to run until it ever try to access layout stuff.

Ok I see. 

You could do something like that using a similar technique to the one sketched above, just using a different logic encapsulated in the mutex-protected state. 

What I like about the above example, is how the concurrent logic can be isolated, versus spread across the event-loop, which is what I think is going to happen if you try to have a global flag and check it at various points. 

Also the nice thing about some state containing the logical flags and a condvar, is that the writer/reader logic can be easily made to fit whatever your business logic needs. If you use an actual reader/writer lock, it's a bit different in the sense that you can't make threads wait for a specific state protected by the lock.

**Comment by asajeffrey** [comment]
@gterzian the problem with trying to do a lock-per-node is that you need to do something to avoid deadlock, e.g. enforce a pattern that before you grab a lock you grab the parent lock, but at that point you may as well just use one global lock.

**Comment by gterzian** [comment]
@asajeffrey Yes that is something to take into account, and I think that if you look again at the example I sketched above, you might agree with what I think is a logical impossibility of deadlock. 

So the pattern you refer to is encoded in the logic of the state protected by the mutex, and the waiting via the condvar which allows threads to wait for that state to be in a specific state, without actually holding the lock(which replaces the need for a parent lock, if I understand what you mean correctly). 

Also, the lock is not held while the read or write operation on the node is happening. It's only held for the short time it takes to read and/or update the mutex protected state. 

So let's say you've got the script-thread reading from a node, and then doing a re-entrant read into that same node. This wouldn't require the mutex to actually be re-entrant, since it's not actually held while reading. The second reader would simply be able to acquire the lock, find `readable` to be true, and increment the `reader_count`. 

Now let's say that while script has two readers, layout wants to write, then layout will acquire the lock, find `writable` to be false, and wait on the condvar. As the two readers drop their reference, the reader count will eventually go to zero, `writeable` go to `true`, and the waiting layout thead will wake-up to find `writeable` to be `true`, and immediately set `readable` to `false` and acquire a mutable reference. 

Now let's say script holds a reference, and then wants to write without first having dropped the reference on the refcell. What you will get is indeed a deadlock in the sense that the thread will wait until `writeable` is `true` forever. 

That could be solved by moving the refcell operations to before the lock one, so that you get the panic that would normally be expected:

```rust
pub fn borrow_mut(&self) -> WrappedRefMut<T> {
         // Ensure you get a "normal" panice instead of a deadlock.
         let ref_mut = self.try_borrow_mut()
            .expect("DomRefCell<T> already borrowed");
        {
            let &(ref lock, ref cvar) = &*self.state; 
            let mut state = lock.lock();
            while !(state.writeable && !state.layout_wants_to_write) {
                // Waiting while holding a local ref mut is not a problem,
                // since we're waiting for the other thread to drop it's writer.
                cvar.wait(&mut state);
            }
            state.readable = false;
            state.writeable = false;
        }  
        WrappedRefMut::new(ref_mut, self.state.clone())
    }
```

Finally, the idea is that a read will most of the time be possible without any waiting, and with `parking_lot` locks being optimized for small size and cheap uncontented operations(both on the mutex and the condvar), you would be adding a tiny bit of overhead, and getting the possibility of full parallelism in return(in so far as nodes can be read from/written to without taking the DOM as a whole into account).

**Comment by gterzian** [comment]
Also, do you want to do this with `DomRefCell`, or `Dom/LayoutDom`? 

let's say `LayoutDom` would look like:

```rust
pub struct LayoutDom<T> {
    state: Arc<(Mutex<DomState>, Condvar)>,
    ptr: ptr::NonNull<T>,
}
```

and `Dom` like:

```rust
#[unrooted_must_root_lint::must_root]
pub struct Dom<T> {
    state: Arc<(Mutex<DomState>, Condvar)>,
    ptr: ptr::NonNull<T>,
}
```

then you can do:

```rust
impl<T: DomObject> LayoutDom<T> {
    /// Returns an unsafe pointer to the interior of this JS object. This is
    /// the only method that be safely accessed from layout. (The fact that
    /// this is unsafe is what necessitates the layout wrappers.)
    pub unsafe fn unsafe_get(&self) -> *const T {
        debug_assert!(thread_state::get().is_layout());
        self.ptr.as_ptr()
    }
}
```

and then you do the condvar thing like above on each access:

```rust

impl<T: DomObject> LayoutDom<T> {
    pub unsafe fn unsafe_get(&self) -> Wrapper<*const T> {
        debug_assert!(thread_state::get().is_layout());
        {
            let &(ref lock, ref cvar) = &*self.state; 
            let mut state = lock.lock();
            while !(state.readable) {
                cvar.wait(&mut state);
            }
            state.writable = false;
            state.reader_counter += 1;
        };
        // As in the earlier example, wrapper will decr the reader counter on drop.
        Wrapper::new(self.ptr.as_ptr(), self.state.clone())
    }
}
```

Then you could even plug this into GC with something like:

```rust
unsafe impl<T: DomObject> JSTraceable for Dom<T> {
    unsafe fn trace(&self, trc: *mut JSTracer) {
        let trace_string;
        let trace_info = if cfg!(debug_assertions) {
            trace_string = format!("for {} on heap", ::std::any::type_name::<T>());
            &trace_string[..]
        } else {
            "for DOM object on heap"
        };
        {
            let &(ref lock, ref cvar) = &*self.state; 
            let mut state = lock.lock();
            while !(state.readable) {
                cvar.wait(&mut state);
            }
            state.writable = false;
            state.reader_counter += 1;
        };

        trace_reflector(trc, trace_info, (*self.ptr.as_ptr()).reflector());

       let &(ref lock, ref cvar) = &*self.state; 
        let mut state = lock.lock();
        state.reader_counter -= 1;
        if state.reader_counter  == 0 {
            state.writeable = true;
            cvar.notify_all();
        }
    }
}
```

**Comment by gterzian** [comment]
Ok so looking at operations like:

```rust
    #[allow(unrooted_must_root)]
    unsafe fn drain_pending_restyles(&self) -> Vec<(LayoutDom<Element>, PendingRestyle)> {
        let mut elements = (*self.unsafe_get())
            .pending_restyles
            .borrow_mut_for_layout();
        // Elements were in a document when they were added to this list, but that
        // may no longer be true when the next layout occurs.
        let result = elements
            .drain()
            .map(|(k, v)| (k.to_layout(), v))
            .filter(|&(ref k, _)| k.upcast::<Node>().get_flag(NodeFlags::IS_CONNECTED))
            .collect();
        result
    }
```

You could also consider moving the locking a notch higher to `Dom/LayouDom`, and perhaps adding a distinction between `unsafe_get` and a new `unsafe_get_mut`(which would acquire a writer and block on the condvar if there are any readers), or just allow only one `unsafe_get` at the time. 

That would also require plugging this mechanism into `Deref for Dom<T>`(and wrapping the return value into something that would drop like the wrappers in the earlier example). 

So I'm not sure if this would be possible at all(not actually familiar with what layout does), but in theory this could allow layout and script to run in parallel, and only block on each other when accessing the same node...

**Comment by gterzian** [comment]
You could also consider moving the locking a notch higher to `Dom` and `LayoutDom`, like:

```rust

pub struct DomState {
     layout_count: usize,
     script_count: usize,
}

#[unrooted_must_root_lint::must_root]
pub struct Dom<T> {
    ptr: ptr::NonNull<T>,
    state: Arc<(Mutex<DomState>, Condvar)>, 
}

impl<T> Dom<T> {
    /// Returns `LayoutDom<T>` containing the same pointer.
    pub unsafe fn to_layout(&self) -> LayoutDom<T> {
        debug_assert!(thread_state::get().is_layout());
        LayoutDom {
            state:: self.state.clone(),
            ptr: self.ptr.clone(),
        }
    }
}

impl<T: DomObject> LayoutDom<T> {
    pub unsafe fn unsafe_get(&self) -> LayoutWrapper<*const T> {
        debug_assert!(thread_state::get().is_layout());
        {
            let &(ref lock, ref cvar) = &*self.state; 
            let mut ref_count = lock.lock();
            while !ref_count.script_count == 0 {
                cvar.wait(&mut state);
            }
           ref_count.layout_count += 1;
        }; 
        // On drop, LayoutWrapper will decrement the count and signal if it hits zero.
         LayoutWrapper::new(self.ptr.as_ptr(), self.state.clone())
    }
}

impl<T: DomObject> Deref for Dom<T> {
    type Target = ScriptWrapper<T>;

    fn deref(&self) -> ScriptWrapper<T> {
        debug_assert!(thread_state::get().is_script());
        {
            let &(ref lock, ref cvar) = &*self.state; 
            let mut ref_count = lock.lock();
            while !ref_count.layout_count == 0 {
                cvar.wait(&mut state);
            }
           ref_count.script_count += 1;
        }; 
       ScriptWrapper::new(self.ptr.as_ptr(), self.state.clone())
    }
}

impl<T: DomObject> Deref for ScriptWrapper<T> {
    type Target = T;

    fn deref(&self) -> &T {
        debug_assert!(thread_state::get().is_script());
        // Not completely sure how all this would integrate with rooting...
        unsafe { &*self.ptr.as_ptr() }
    }
}

```

**Comment by asajeffrey** [comment]
What happens if we have two nodes, A and B, with the access pattern:
- script accesses node B, sets B's script count to 1,
- layout accesses node A, sets A's layout count to 1,
- script tries to access node A, waits on A's condvar for A's layout count to be 0, and
- layout tries to access node B, waits on B's condvar for B' script count to be 0.

Isn't this going to produce deadlock?

Also, isn't there scope for inconsistency here, e.g. A and B are both red, then script sets B green them A green, coudn't layout have a transitory state when A is green and B is red?

**Comment by nox** [comment]
The point of not using a fine-grained locking strategy is that once layout starts then it never has to wait. It just sees a tree of glorified shared references, while script can continue its own business until it reaches a DOM call that wants access to anything that is used by layout.

**Comment by nox** [comment]
> I think this would also mean ensuring we never leak `DomRoot` somewhere, since that could make layout block forever...

Just possessing a `DomRoot` should never block anything: `DomRoot` never lets you mutate it contents. Mutation is the only thing that can break layout and it only happens through types that provide inner mutability such as `DomRefCell<T>`.

**Comment by gterzian** [comment]
> Isn't this going to produce deadlock?

Yes that scenario would indeed be a deadlock. 

And this scenario might in practice be mitigated by the tree structure of the DOM? In the sense that you would grab references to node in order of nesting, and the other thread wouldn't be able to grab a ref to a nested node ahead of the other thread who is traversing?

> Also, isn't there scope for inconsistency here, e.g. A and B are both red, then script sets B green them A green, coudn't layout have a transitory state when A is green and B is red?

I think the locking would prevent inconsistencies, however perhaps you're talking about layout storing the result of it's work somewhere else (display list?), and this then later being invalidated by the script-thread doing a mutation?

That's sort of what I meant with "is this useful at all", in the sense that having those threads run in parallel might actually not produce useful results and blocking one of the threads might be better...

> The point of not using a fine-grained locking strategy is that once layout starts then it never has to wait. It just sees a tree of glorified shared references, while script can continue its own business until it reaches a DOM call that wants access to anything that is used by layout.

Ok, so here when you write "access to anything", do you mean "mutable access" as per your other comment?

Do I understand it correctly that layout basically starts at https://github.com/servo/servo/blob/9fb83d81438d425e06d061f8ba0c4128c61fc1ac/components/script/dom/window.rs#L1629 

In that case, why not, instead of subsequently blocking on the join channel:

1. Store the channel in a thread local `Option<Sender<ReflowComplete>>`.
2. On each `borrow_mut` on a `DomRefCell`, check if it is `Some`. 
3. If so, block on it, and run the image cache and `ScriptThread::note_newly_transitioning_nodes`
4. And also, at the end of a turn of the event-loop, also check if it is some(in case there would have been no mutable access to anything) and block on it. Probably at the similar point were currently reflows are batched. 

I think one thing that's tricky is the "noting of transitioning nodes" and handling of transition event messages from the constellation. I think there is an opportunity to miss transition events if you haven't blocked on the reflow result yet.

Currently it looks like calling `window.reflow` doesn't always happen inside a task, because currently the script-event loop doesn't really run one task at a time. For example UI events are fired straight when handling a constellation message. 

So what happens if the script-thread calls `window.reflow`, but instead of blocking on the result from layout, and noting the newly transitioning nodes, you don't block until the next call to `borrow_mut`, and then instead of first running a task, you first handle a `ConstellationControlMsg::TransitionEnd`, but you haven't noted the nodes yet? 

Well I guess the answer is to simply have the script-thread check if there is a channel from layout to block on at the start of `handle_transition_event`, and block on it if necessary before actually running the transition logic...

**Comment by gterzian** [comment]
Also one other thing, I wasn't aware of the exact usecase for `DomRefCell`, perhaps I wasn't the only one, so I've been using it basically everytime I needed a `RefCell` inside a `dom_struct`, even if the thing inside of it isn't actually related to the DOM(things like `DomRefCell<Vec<u8>>`). 

So if the goal is to have script run a task "as far as possible" and not block until hitting a potential DOM mutation, we might also want to be more consistent in the distinction between `RefCell` and `DomRefCell`, and only use the latter for values related to the DOM?

**Comment by asajeffrey** [comment]
> Yes that scenario would indeed be a deadlock. Maybe the tree structure of the DOM could mitigate that scenario in practice?

I don't think so, the only way that would be true is if we maintained an invariant that before locking a node, you locked all its ancestors, but this would essentially be the same as one global lock (the root element's lock).

> In that case, why not, instead of subsequently blocking on the join channel:
> 
>     1. Store the channel in a thread local `HashMap<PipelineId, Sender<ReflowComplete>>`.
> 
>     2. On each `borrow_mut` on a `DomRefCell`, check if there is a channel to block on.
> 
>     3. If so, block on it, and run the image cache and `ScriptThread::note_newly_transitioning_nodes`
> 
>     4. And also, at the end of a turn of the event-loop, also check for each active pipeline if there is a channel to block on(in case there would have been no mutable access to anything) and block on it. Probably at the similar point were currently reflows are batched. Probably also drop any channels in the map for not fully active pipelines.

Thus would probably work, but seems fiddlier than just using a mutex.

**Comment by asajeffrey** [comment]
@nox is it possible/worth it to split `DOMRefCell` into two types, one that's only used by script, and one that's accessible from layout? (I am a bit ignorant about layout_2020, you may already be doing this.)

**Comment by nox** [comment]
Yes. Just start using `RefCell` again for non-layout stuff.

I don't understand why we would need additional thread local hash maps and whatnot.
